ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -1,123 +1,348 @@
+        let quotient = UInt::new(|| "quotient", cb)?;
+        let remainder = UInt::new(|| "remainder", cb)?;
+
+        // `rem_e` and `div_e` are expressions representing the remainder and
+        // divisor from the signed or unsigned division operation, with signs
+        // normalized to be nonnegative, so that correct values must satisfy
+        // either `0 <= rem_e < div_e` or `div_e == 0`.  The `rem_e` value
+        // should be constrained to be nonnegative before being returned from
+        // this block, while the checks `rem_e < div_e` or `div_e == 0` are
+        // done later.
+        let (internal_config, rem_e, div_e) = match I::INST_KIND {
+            InsnKind::DIVU | InsnKind::REMU => {
+                cb.require_equal(
+                    || "unsigned_division_relation",
+                    dividend.value(),
+                    divisor.value() * quotient.value() + remainder.value(),
+                )?;
+
+                (InternalDivRem::Unsigned, remainder.value(), divisor.value())
+            }
+
+            InsnKind::DIV | InsnKind::REM => {
+                let dividend_signed =
+                    Signed::construct_circuit(cb, || "dividend_signed", &dividend)?;
+                let divisor_signed = Signed::construct_circuit(cb, || "divisor_signed", &divisor)?;
+                let quotient_signed =
+                    Signed::construct_circuit(cb, || "quotient_signed", &quotient)?;
+                let remainder_signed =
+                    Signed::construct_circuit(cb, || "remainder_signed", &remainder)?;
+
+                // The quotient and remainder can be interpreted as non-positive
+                // values when exactly one of dividend and divisor is negative
+                let neg_div_expr = {
+                    let a_neg = dividend_signed.is_negative.expr();
+                    let b_neg = divisor_signed.is_negative.expr();
+                    // a_neg xor b_neg
+                    &a_neg + &b_neg - 2 * a_neg * b_neg
+                };
+                let negative_division = cb.flatten_expr(|| "neg_division", neg_div_expr)?;
+
+                // Check for signed division overflow: i32::MIN / -1
+                let is_dividend_max_negative = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_dividend_max_negative",
+                    dividend.value(),
+                    i32::MIN.into(),
+                )?;
+                let is_divisor_minus_one = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_divisor_minus_one",
+                    divisor.value(),
+                    (-1i32).into(),
+                )?;
+                let is_signed_overflow = cb.flatten_expr(
+                    || "signed_division_overflow",
+                    is_dividend_max_negative.expr() * is_divisor_minus_one.expr(),
+                )?;
+
+                // For signed division overflow, dividend = -2^31 and divisor
```

Comment:
I suspect we actually might be able to get away without any special handling of signed division overflow (and just get the right answer out of the general constraints).

But that's something for me to think about later, nothing that should hold up this PR.

> My suspicion is because we have more range available in Goldilocks, and (i32::MIN as Goldilocks) / (-1) == 1<<31 as Goldilocks, which is the answer we want at the end.

Reply:
I think the key point here is not Goldilocks field range, it's riscv signed 32 bit registet can not represent overflow 2^31. So riscv define a special case for it.

Reply:
I know.  I am saying that the special case disappears in our setting!